dashboard/modules/job/tests/test_sdk.py

```diff
@@ -142,5 +154,88 @@ def test_temporary_uri_reference(monkeypatch, expiration_s):
+    "ray_start_cluster_head", [{"include_dashboard": True}], indirect=True
+)
+def test_job_head_choose_job_agent_E2E(mock_candidate_number, ray_start_cluster_head):
+    cluster = ray_start_cluster_head
+    assert wait_until_server_available(cluster.webui_url) is True
+    webui_url = cluster.webui_url
+    webui_url = format_web_url(webui_url)
+    client = JobSubmissionClient(webui_url)
+
+    def get_register_agents_number():
+        response = requests.get(webui_url + "/internal/node_module")
+        response.raise_for_status()
+        result = response.json()
+        data = result["data"]
+        return data["registeredAgents"]
+
+    def submit_job_and_wait_finish():
+        submission_id = client.submit_job(entrypoint="echo hello")
+
+        wait_for_condition(_check_job_succeeded, client=client, job_id=submission_id)
+
+    # make sure list(cluster.worker_nodes)[0] will be the owner of a supervisor actor.
+    cluster.add_node(dashboard_agent_listen_port=52366)
+    wait_for_condition(lambda: get_register_agents_number() == 2, timeout=20)
+    assert len(cluster.worker_nodes) == 1
+    node_try_to_kill = list(cluster.worker_nodes)[0]
+    submit_job_and_wait_finish()
+
+    cluster.add_node(dashboard_agent_listen_port=52367)
+    wait_for_condition(lambda: get_register_agents_number() == 3, timeout=20)
+
+    submit_job_and_wait_finish()
+    submit_job_and_wait_finish()
+    submit_job_and_wait_finish()
+
+    def get_all_new_supervisor_actor_info(old_supervisor_actor):
+        all_actors = ray.state.state.actor_table(None)
+        res = dict()
+        for actor_id, actor_info in all_actors.items():
+            if actor_id in old_supervisor_actor:
+                continue
+            if not actor_info["Name"].startswith("_ray_internal_job_actor"):
+                continue
+            res[actor_id] = actor_info
+        return res
+
+    old_supervisor_actor = set()
+    new_supervisor_actor = get_all_new_supervisor_actor_info(old_supervisor_actor)
+    new_owner_port = set()
+    for actor_id, actor_info in new_supervisor_actor.items():
+        old_supervisor_actor.add(actor_id)
+        new_owner_port.add(actor_info["OwnerAddress"]["Port"])
+
+    assert len(new_owner_port) == 2
+    old_owner_port = new_owner_port
+
+    node_try_to_kill.kill_raylet()
+
+    # make sure the head updates the info of the dead node.
+    wait_for_condition(lambda: get_register_agents_number() == 2, timeout=20)
+
+    submit_job_and_wait_finish()
+    submit_job_and_wait_finish()
+    submit_job_and_wait_finish()
+
+    new_supervisor_actor = get_all_new_supervisor_actor_info(old_supervisor_actor)
+    new_owner_port = set()
+    for actor_id, actor_info in new_supervisor_actor.items():
+        old_supervisor_actor.add(actor_id)
+        new_owner_port.add(actor_info["OwnerAddress"]["Port"])
+    assert len(new_owner_port) == 2
+    assert len(old_owner_port - new_owner_port) == 1
+    assert len(new_owner_port - old_owner_port) == 1
```

Comment:
This test is observed to be flaky in two different ways, it's easily reproducible locally:
```
>       assert len(new_owner_port) == 2
--
  | E       assert 1 == 2
  | E         +1
  | E         -2
```

```
>       assert len(old_owner_port - new_owner_port) == 1
--
  | E       assert 0 == 1
  | E         +0
  | E         -1
```

Are we using "Port" as a kind of Node ID here?  And are we checking that after the raylet for the first worker node (`dashboard_agent_listen_port=52366`) gets killed, the next three job supervisors are started up on both the second worker node(`dashboard_agent_listen_port=52367`) and on the head node?  If so, could we have a bad outcome due to chance?

Here's one print output from a local test run:
```
old_owner_port {60321, 60407}
new_owner_port {60407}
```
Perhaps the old jobs were on the head node and the second worker node, and the new jobs were all on the second worker node?

Reply:
@architkulkarni There are indeed some problems with the implementation of this place. I wrote it again. could you take look at it again?